linalg/cudss.hpp

```diff
@@ -0,0 +1,181 @@
+      LOWER = 1,
+      /// CUDSS_MVIEW_UPPER: Upper-triangular matrix (including the diagonal) All values below the main diagonal will be ignored.
+      UPPER = 2,
+   };
+
+   CuDSSSolver(MPI_Comm comm);
+
+   /**
+    * @brief Set the matrix type
+    *
+    * Supported matrix types:
+    *          CuDSSSolver::UNSYMMETRIC,
+    *          CuDSSSolver::SYMMETRIC_INDEFINITE,
+    *      and CuDSSSolver::SYMMETRIC_INDEFINITE,
+    *
+    * @param mtype_ Matrix type
+    *
+    * @note This method has to be called before SetOperator
+    */
+   void SetMatrixSymType(MatType mtype_);
+
+   /**
+    * @brief Set the matrix view type
+    *
+    * Supported matrix types:
+    *          CuDSSSolver::FULL,
+    *          CuDSSSolver::LOWER,
+    *      and CuDSSSolver::UPPER,
+    *
+    * @param mvtype Matrix view type
+    *
+    * @note This method has to be called before SetOperator
+    */
+   void SetMatrixViewType(MatViewType mvtype);
+
+   /**
+    * @brief Set the flag controlling reuse of the symbolic factorization
+    * for multiple operators
+    *
+    * @param reuse Flag to reuse symbolic factorization
+    *
+    * @note This method has to be called before repeated calls to SetOperator
+    */
+   void SetReorderingReuse(bool reuse);
+
+   /**
+    * @brief Set the flag controlling sort the rows of CSR matrix
+    *
+    * @param sort_row_ Flag to sort the row of CSR matrix
+    *
+    * @note This method has to be called before repeated calls to SetOperator
+    */
+   void SetMatrixSortRow(bool sort_row_);
+
+   void SetOperator(const Operator &op) override;
+   /**
+    * @brief Solve $ y = Op^{-1} x $
+    *
+    * @param x RHS vector
+    * @param y Solution vector
+    */
+   void Mult(const Vector &x, Vector &y) const override;
+
+   /**
+    * @brief Solve $ Y_i = Op^{-1} X_i $
+    *
+    * @param X Array of RHS vectors
+    * @param Y Array of Solution vectors
+    */
+   void ArrayMult(const Array<const Vector *> &X, Array<Vector *> &Y) const;
+
+   ~CuDSSSolver();
+
+private:
+   // MPI_Comm
+   MPI_Comm *mpi_comm = NULL;
+   // MPI rank
+   int myid;
+
+   // Parameter controlling whether or not to reuse the symbolic factorization
+   // for multiple calls to SetOperator
+   bool reorder_reuse = false;
+
+   // Parameter controlling whether or not to sort the row of CSR matrix
+   bool sort_row = false;
+
+   // Matrix type
+   MatType mtype = MatType::UNSYMMETRIC;
+   // Parameter controlling the matrix type
+   cudssMatrixType_t mat_type = CUDSS_MTYPE_GENERAL;
+
+   mutable int nrhs = 0; // the number of the RHSs
+   int n = 0; // global number of rows
+
+   int nnz = 0; // the number of non zeros
+   int row_start = 0;  // the first row index in CSR matrix operator
+   int row_end = 0;   // the end row index in CSR matrix operator
+   int n_loc = 0;    // the number of the rows in CSR matrix operator
```

Comment:
This class is derived from `Solver`, which is derived from `Operator`, which has the member `height`. Is `n_loc` the same as `height`?